Fix Issue #131: S3 Metadata

PopeFelix opened the pull request:
This addresses [issue 131](https://github.com/pplu/aws-sdk-perl/issues/131) by adding a Trait, "ParamsInHeader" for free-form AWS parameters that are passed in the request header (e.g. S3 Metadata). The REST XML request builder is updated to look for that trait and act accordingly. 

Testing done: I created a sketch with my company's AWS credentials and was able to put an object with S3 metadata specified, like so:

```
    use 5.016;
    use FindBin qw($Bin);
    use lib qq($Bin/../auto-lib);
    use lib qq{$Bin/../lib};
    use Paws;

    # AWS credentials in %ENV
    my $s3 = Paws->service('S3', region => 'us-east-1');
    my $bucket = 'mah-bukkit';
    my $key = 'Paws-test-foobar';
    my $data = 'my data';

    $s3->PutObject( Bucket => $bucket, Key => $key, Body => $data, Metadata => { filename => 'foobar.doc' } );
    my $object = $s3->GetObject( Bucket => $bucket, Key => $key );
    use DDP;
    p $object;
    say 'metadata:';
    p $object->Metadata;
```

```
Paws::S3::GetObjectOutput  {
    Parents       Moose::Object
    public methods (29) : AcceptRanges, Body, CacheControl, ContentDisposition, ContentEncoding, ContentLanguage, ContentLength, ContentRange, ContentType, DeleteMarker, ETag, Expiration, Expires, LastModified, meta, Metadata, MissingMeta, PartsCount, ReplicationStatus, RequestCharged, Restore, ServerSideEncryption, SSECustomerAlgorithm, SSECustomerKeyMD5, SSEKMSKeyId, StorageClass, TagCount, VersionId, WebsiteRedirectLocation
    private methods (2) : _request_id, _stream_param
    internals: {
        _request_id     "2DD05179C822C868",
        AcceptRanges    "bytes",
        Body            "my data",
        ContentLength   7,
        ContentType     "application/octet-stream",
        ETag            ""1291e1c0aa879147f51f4a279e7c2e55"",
        LastModified    "Sat, 04 Mar 2017 05:37:08 GMT",
        Metadata        Paws::S3::Metadata
    }
}
metadata:
Paws::S3::Metadata  {
    Parents       Moose::Object
    public methods (4) : Map, meta, xml_keys, xml_values
    private methods (0)
    internals: {
        Map   {
            x-amz-meta-filename   "foobar.doc"
        }
    }
}
```

PopeFelix commented:
I'm happy with this now, so feel free to merge. ;)

pplu commented:
I'm merging the code into a "staging" branch first. I want to get some tests for this (so it doesn't break in the future).

PopeFelix commented:
Makes sense to me. ☺

On Apr 3, 2017 18:32, "Jose Luis Martinez" <notifications@github.com> wrote:

I'm merging the code into a "staging" branch first. I want to get some
tests for this (so it doesn't break in the future).

—
You are receiving this because you authored the thread.
Reply to this email directly, view it on GitHub
<https://github.com/pplu/aws-sdk-perl/pull/152#issuecomment-291317212>, or mute
the thread
<https://github.com/notifications/unsubscribe-auth/AAJqGHO_zleD30Y3kxS6Xi03_RALxzvFks5rsYGYgaJpZM4MSdAr>
.

pplu commented:
Hi!

  The branch is almost ready to merge: https://github.com/pplu/aws-sdk-perl/compare/fix/s3-metadata

   I've introduced a small behaviour change with respect to your proposal: the hash being returned by GetObject was of the form `{ 'x-amz-meta-KEY' => 'VALUE' }`. I've changed that so that GetObject returns `{ 'KEY' => 'VALUE' }` (so that the fact that the metadata comes from headers with a prefix is hidden from the user.

   If you're OK with that, I can merge, and prepare for release

PopeFelix commented:
I think that's fine.

KP

On Tue, Apr 4, 2017 at 5:43 PM, Jose Luis Martinez <notifications@github.com
> wrote:

> Hi!
>
> The branch is almost ready to merge: https://github.com/pplu/aws-
> sdk-perl/compare/fix/s3-metadata
>
> I've introduced a small behaviour change with respect to your proposal:
> the hash being returned by GetObject was of the form { 'x-amz-meta-KEY'
> => 'VALUE' }. I've changed that so that GetObject returns { 'KEY' =>
> 'VALUE' } (so that the fact that the metadata comes from headers with a
> prefix is hidden from the user.
>
> If you're OK with that, I can merge, and prepare for release
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/pplu/aws-sdk-perl/pull/152#issuecomment-291659639>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAJqGNhsZynit-e2YSOSfdQczkcSA4O8ks5rsseqgaJpZM4MSdAr>
> .
>



-- 
Kit Peters, W0KEH
GPG public key fingerpint: 1A12 04B6 0C80 306A B292  14FD 2C7A 1037 F666
46A7
Hello to any and all NSA, DEA, or other government or non-government agents
reading this email.  Tell me about your life; I'll tell you about mine.

tryorfry commented:
Hi,
I wish once Metadata is attached in PutObject, it's available in all S3 methods for example in ListAllObjects. Looks like it's available only in GetObject. 
Am I missing something?

pplu commented:
Hi,

  ListObjects and ListObjectsV2 don't return the metadata of the objects, so Paws here is limited to what is returned by S3 APIs (ListAllObjects is just a wrapper around ListObjects to iterate over the results for you). You would have to do a $s3->HeadObject or GetObject on each object returned by the list to get the metadata. I think HeadObject is returning metadata successfully.

ListObjects docu:
https://docs.aws.amazon.com/AmazonS3/latest/API/RESTBucketGET.html
https://docs.aws.amazon.com/AmazonS3/latest/API/v2-RESTBucketGET.html

HeadObject
https://docs.aws.amazon.com/AmazonS3/latest/API/RESTObjectHEAD.html

tryorfry commented:
Thanks Jose again.

On 9 Feb 2018 17:07, "Jose Luis Martinez" <notifications@github.com> wrote:

Hi,

ListObjects and ListObjectsV2 don't return the metadata of the objects, so
Paws here is limited to what is returned by S3 APIs (ListAllObjects is just
a wrapper around ListObjects to iterate over the results for you). You
would have to do a $s3->HeadObject or GetObject on each object returned by
the list to get the metadata. I think HeadObject is returning metadata
successfully.

ListObjects docu:
https://docs.aws.amazon.com/AmazonS3/latest/API/RESTBucketGET.html
https://docs.aws.amazon.com/AmazonS3/latest/API/v2-RESTBucketGET.html

HeadObject
https://docs.aws.amazon.com/AmazonS3/latest/API/RESTObjectHEAD.html

—
You are receiving this because you commented.
Reply to this email directly, view it on GitHub
<https://github.com/pplu/aws-sdk-perl/pull/152#issuecomment-364376388>, or mute
the thread
<https://github.com/notifications/unsubscribe-auth/ADxS_pMKmEo8iBkF7VzvDZY-t_zym_Vsks5tTArtgaJpZM4MSdAr>
.
